File: surf-api-bukkit/surf-api-bukkit-server/src/main/java/dev/slne/surf/surfapi/bukkit/server/impl/nms/bridges/packets/PacketOperationImpl.java
package dev.slne.surf.surfapi.bukkit.server.impl.nms.bridges.packets;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

import com.google.common.flogger.FluentLogger;
import com.google.common.flogger.StackSize;
import dev.slne.surf.surfapi.bukkit.api.nms.bridges.packets.PacketOperation;
import dev.slne.surf.surfapi.bukkit.server.nms.NmsUtil;
import java.util.LinkedList;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;
import net.minecraft.network.protocol.Packet;
import net.minecraft.network.protocol.game.ClientGamePacketListener;
import net.minecraft.network.protocol.game.ClientboundBundlePacket;
import org.bukkit.entity.Player;

public final class PacketOperationImpl implements PacketOperation, NmsUtil {

  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  private Operation operation;

  private PacketOperationImpl(Operation operation) {
    this.operation = operation;
  }

  private PacketOperationImpl() {
    this.operation = Operation.empty();
  }

  @Override
  public void execute(Player player) {
    final var connection = toNms(player).connection;
    final var packets = operation.apply(player, new LinkedList<>());

    if (packets.isEmpty()) {
      logger.atInfo()
          .atMostEvery(60, TimeUnit.SECONDS)
          .withStackTrace(StackSize.SMALL)
          .log("No packets to send for player '%s'. Is this intended behaviour?", player.getName());
      return;
    }

    if (packets.size() == 1) {
      connection.send(packets.getFirst());
      return;
    }

    connection.send(new ClientboundBundlePacket(packets));
  }

  @Override
  public PacketOperationImpl add(PacketOperation other) {
    checkArgument(other instanceof PacketOperationImpl,
        "operation must be an instance of PacketOperationImpl");

    this.operation = this.operation.andThen(((PacketOperationImpl) other).operation);
    return this;
  }

  public static PacketOperationImpl empty() {
    return new PacketOperationImpl();
  }

  public static PacketOperationImpl complex(Operation operation) {
    return new PacketOperationImpl(operation);
  }

  public static PacketOperationImpl simple(Function<Player, Packet<? super ClientGamePacketListener>> packetSupplier) {
    return new PacketOperationImpl((player, packets) -> {
      packets.add(packetSupplier.apply(player));
      return packets;
    });
  }

  @FunctionalInterface
  public interface Operation {

    LinkedList<Packet<? super ClientGamePacketListener>> apply(Player player,
        LinkedList<Packet<? super ClientGamePacketListener>> packets);

    default Operation andThen(Operation after) {
      checkNotNull(after, "after");

      return (player, packets) -> after.apply(player, apply(player, packets));
    }

    static Operation empty() {
      return (player, packets) -> packets;
    }
  }
}

File: surf-api-bukkit/surf-api-bukkit-server/src/main/kotlin/dev/slne/surf/surfapi/bukkit/server/impl/nms/bridges/packets/PacketOperationImpl.kt
package dev.slne.surf.surfapi.bukkit.server.impl.nms.bridges.packets

import com.google.common.flogger.FluentLogger
import dev.slne.surf.surfapi.bukkit.api.nms.bridges.packets.PacketOperation
import dev.slne.surf.surfapi.bukkit.server.nms.toNms
import net.minecraft.network.protocol.Packet
import net.minecraft.network.protocol.game.ClientGamePacketListener
import net.minecraft.network.protocol.game.ClientboundBundlePacket
import org.bukkit.entity.Player
import java.util.*

class PacketOperationImpl : PacketOperation {
    private var operation: Operation

    private constructor(operation: Operation) {
        this.operation = operation
    }

    private constructor() {
        this.operation = Operation.empty()
    }

    override fun execute(player: Player) {
        val connection = player.toNms().connection
        val packets = operation.apply(
            player,
            LinkedList<Packet<in ClientGamePacketListener>>()
        )

        if (packets.isEmpty()) {
            return
        }

        if (packets.size == 1) {
            connection.send(packets.first())
            return
        }

        connection.send(ClientboundBundlePacket(packets))
    }

    override fun add(other: PacketOperation): PacketOperationImpl {
        require(other is PacketOperationImpl) { "operation must be an instance of PacketOperationImpl" }

        this.operation = this.operation.andThen(other.operation)
        return this
    }

    override fun isEmpty(): Boolean {
        val operation = operation
        return operation is EmptyOperation && operation.empty
    }

    fun interface Operation {
        fun apply(
            player: Player,
            packets: LinkedList<Packet<in ClientGamePacketListener>>
        ): LinkedList<Packet<in ClientGamePacketListener>>

        fun andThen(after: Operation): Operation {
            return Operation { player, packets ->
                after.apply(player, apply(player, packets))
            }
        }

        companion object {
            fun empty() = EmptyOperation()
        }
    }

    class EmptyOperation: Operation {
        var empty: Boolean = true
            private set

        override fun apply(
            player: Player,
            packets: LinkedList<Packet<in ClientGamePacketListener>>
        ): LinkedList<Packet<in ClientGamePacketListener>> {
            return packets
        }

        override fun andThen(after: Operation): Operation {
            empty = false
            return super.andThen(after)
        }
    }

    companion object {
        private val logger: FluentLogger = FluentLogger.forEnclosingClass()

        @JvmStatic
        fun empty(): PacketOperationImpl {
            return PacketOperationImpl()
        }

        @JvmStatic
        fun complex(operation: Operation): PacketOperationImpl {
            return PacketOperationImpl(operation)
        }

        fun simple(packetSupplier: (Player) -> Packet<in ClientGamePacketListener>): PacketOperationImpl {
            return PacketOperationImpl { player, packets ->
                packets.add(packetSupplier(player))
                packets
            }
        }
    }
}
